luigi/contrib/lsf.py

```diff
@@ -0,0 +1,354 @@
+and https://wiki.med.harvard.edu/Orchestra/IntroductionToLSF
+
+See: https://github.com/spotify/luigi/issues/1936
+
+This extension is modeled after the hadoop.py approach.
+I'll be making a few assumptions, and will try to note them.
+
+Going into it, the assumptions are:
+
+- You schedule your jobs on an LSF submission node.
+- The 'bjobs' command on an LSF batch submission system returns a standardized format.
+- All nodes have access to the code you're running.
+- The sysadmin won't get pissed if we run a 'bjobs' check every thirty
+  seconds or so per job (there are ways of coalescing the bjobs calls if that's not cool).
+
+The procedure:
+
+- Pickle the class
+- Construct a bsub argument that runs a generic runner function with the path to the pickled class
+- Runner function loads the class from pickle
+- Runner function hits the work button on it
+
+"""
+
+LOGGER = logging.getLogger('luigi-interface')
+
+
+def track_job(job_id):
+    """
+    Tracking is done by requesting each job and then searching for whether the job
+    has one of the following states:
+    - "RUN",
+    - "PEND",
+    - "SSUSP",
+    - "EXIT"
+    based on the LSF documentation
+    """
+    cmd = "bjobs -noheader -o stat %d" % job_id
+    track_job_proc = subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, shell=True)
+    status = track_job_proc.communicate()[0].strip('\n')
+    return status
+
+
+def kill_job(job_id):
+    """
+    Kill a running LSF job
+    """
+    subprocess.call(['bkill', job_id])
+
+
+class LSFJobTask(luigi.Task):
+    """
+    Takes care of uploading and executing an LSF job
+    """
+
+    n_cpu_flag = luigi.IntParameter(default=2, significant=False)
+    shared_tmp_dir = luigi.Parameter(default='/tmp', significant=False)
+    resource_flag = luigi.Parameter(default='mem=8192', significant=False)
+    memory_flag = luigi.Parameter(default='8192', significant=False)
+    queue_flag = luigi.Parameter(default='queue_name', significant=False)
+    runtime_flag = luigi.IntParameter(default=60)
+    job_name_flag = luigi.Parameter(default='')
+    poll_time = luigi.FloatParameter(
+        significant=False, default=5,
+        description="specify the wait time to poll bjobs for the job status")
+    save_job_info = luigi.BoolParameter(default=False)
+    output = luigi.Parameter(default='')
+    extra_bsub_args = luigi.Parameter(default='')
+
+    job_status = None
+
+    def fetch_task_failures(self):
+        """
+        Read in the error file from bsub
+        """
+        error_file = os.path.join(self.tmp_dir, "job.err")
+        if os.path.isfile(error_file):
+            with open(error_file, "r") as f_err:
+                errors = f_err.readlines()
+        else:
+            errors = ''
+        return errors
+
+    def fetch_task_output(self):
+        """
+        Read in the output file
+        """
+        # Read in the output file
+        with open(os.path.join(self.tmp_dir, "job.out"), "r") as f_out:
```

Comment:
Nitpicky, but with errors you check if the file exists, is there a reason the same logic isn't mirrored here?